sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1162 @@
+value to determine which of the extension's APIs the implementation supports.
+
+Table 1. Values of the `SYCL_EXT_ONEAPI_GRAPH` macro.
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version. Base features are supported.
+|===
+
+=== SYCL Graph Terminology
+
+:explicit-memory-ops: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:explicitmemory
+
+Table 2. Terminology.
+[%header,cols="1,3"]
+|===
+| Concept | Description
+
+| Graph
+| A directed and acyclic graph (DAG) of commands (nodes) and their dependencies
+(edges), represented by the `command_graph` class.
+
+| Node
+| A command, which can have different attributes.
+
+| Edge
+| Dependency between commands as a happens-before relationship.
+
+|===
+
+==== Explicit Graph Building API
+
+When using the explicit graph building API to construct a graph, nodes and
+edges are captured as follows.
+
+Table 3. Explicit Graph Definition.
+[%header,cols="1,3"]
+|===
+| Concept | Description
+
+| Node
+| In the explicit graph building API nodes are created by the user invoking
+methods on a modifiable graph. Each node represent either a command-group
+function, empty operation, or device memory allocation/free.
+
+| Edge
+| In the explicit graph building API edges are defined by the user. This is
+either through buffer accessors, the `make_edge()` function, or by passing
+dependent nodes on creation of a new node.
+|===
+
+==== Queue Recording API
+
+When using the record & replay API to construct a graph by recording a queue,
+nodes and edges are captured as follows.
+
+Table 4. Recorded Graph Definition.
+[%header,cols="1,3"]
+|===
+| Concept | Description
+
+| Node
+| Nodes in a queue recorded graph represent each of the command group
+submissions of the program. Each submission encompasses either one or both of
+a.) some data movement, b.) a single asynchronous kernel launch. Nodes cannot
+define forward edges, only backwards. This is, kernels can only create
+dependencies on command-groups that have already been submitted. This means that
+transparently a node can depend on a previously recorded graph (sub-graph),
+which works by creating edges to the individual nodes in the old graph. Explicit
+memory operations without kernels, such as a memory copy, are still classed as
+nodes under this definition, as the
+{explicit-memory-ops}[SYCL 2020 specification states] that these can be seen as
+specialized kernels executing on the device.
+
+| Edge
+| An edge in a queue recorded graph represents a data dependency between two
+nodes. Data dependencies can naturally be expressed in user code through buffer
+accessors. USM pointers also convey data dependencies, however offsets into
+system allocations (`malloc`/`new`) are not supported.
+|===
+
+==== Sub-Graph
+
+A node in a graph can take the form of a nested sub-graph. This occurs when
+a command-group submission that invokes `handler::exec_graph()` with an
```

Comment:
I don't think there are any other instances of "exec" being used this way in SYCL.

`handler::submit_graph` would be more consistent with the naming of `submit`... But then it would be the only `queue` shortcut with "submit" in the name.  I'm wondering if this should be called something like `graph`, in the same way that `parallel_for` is not called `execute_parallel_for`?

Reply:
The rationale for `parallel_for`/`execute_parallel_for` makes sense, will tweak this to `handler::graph()`